### ISSUE TITLE
Constant gets evaluated even when it contains invalid `transmute`

### ISSUE THREAD

**Issue by RalfJung** [rust-lang/rust#79047]
Consider the following code:
```rust
const DST: &[u8] = unsafe { std::mem::transmute(1usize) };

fn main() {
    match &b""[..] {
        DST => {}
    }
}
```
This code hits a peculiar code path in the interpreter:

https://github.com/rust-lang/rust/blob/30e49a9ead550551e879af64ba91a0316da1c422/compiler/rustc_mir/src/interpret/place.rs#L916-L927

This code path *should* be unreachable: just like the interpreter can assume that code code it runs on is well-typed, it should be able to assume that the code it runs on has its transmutes checked. But something seems to be different about `transmute` when compared with "normal" type-checking.

`TransmuteSizeDiff` is a hack; we should instead arrange things in a way that failing the transmute check inhibits const-evaluation the same way that `true + 4` inhibitis const-evaluation.

**Comment by RalfJung** [comment]
The same issue also arises with `SizeOf` and its requirement that the type be sized.
Cc https://github.com/rust-lang/rust/pull/81185#issuecomment-765217308

**Comment by osa1** [comment]
> The same issue also arises with SizeOf and its requirement that the type be sized.

Is it really the same issue? My understanding is the `SizeOf` issue is a type error (as the argument's `Sized` bound is not satisfied), but the transmute error is caught not by the type checker, but rather it has its own check in

https://github.com/rust-lang/rust/blob/85e355ea9bd86ac6580a5d422a65dbf689845808/compiler/rustc_passes/src/intrinsicck.rs#L71-L124

So I think it's possible that we have two different bugs here.

**Comment by RalfJung** [comment]
Well, maybe not exactly the same issue, but something very similar at least.
